File: FinancialBI/Views/PeriodRangeSelectionControl.vb
Imports System.Collections.Generic

Public Class PeriodRangeSelectionControl

#Region "New"

    Friend Sub New(ByRef p_versionId As UInt32)

        ' This call is required by the designer.
        InitializeComponent()

        ' Add any initialization after the InitializeComponent() call.
        MultilanguageSetup()
        PeriodsRangeSetup(p_versionId)

        AddHandler m_startDate.ValueChanged, AddressOf m_startDate_ValueChanged
        AddHandler m_endDate.ValueChanged, AddressOf m_endDate_ValueChanged

    End Sub

    Private Sub MultilanguageSetup()

        m_startDateLabel.Text = Local.GetValue("submissionsFollowUp.start_date")
        m_endDateLabel.Text = Local.GetValue("submissionsFollowUp.end_date")

    End Sub

    Private Sub PeriodsRangeSetup(ByRef p_versionId As UInt32)

        m_startDate.FormatValue = "MM-dd-yy"
        m_endDate.FormatValue = "MM-dd-yy"

        Dim l_periods As Int32() = GlobalVariables.Versions.GetPeriodsList(p_versionId)
        If l_periods Is Nothing OrElse l_periods.Length = 0 Then Exit Sub

        m_startDate.MinDate = Date.FromOADate(l_periods(0))
        m_startDate.MaxDate = Date.FromOADate(l_periods(l_periods.Length - 1))
        m_endDate.MinDate = Date.FromOADate(l_periods(0))
        m_endDate.MaxDate = Date.FromOADate(l_periods(l_periods.Length - 1))

        Dim l_startDate As Date = Today.Date
        l_startDate = l_startDate.AddDays(-Period.m_nbDaysInWeek)

        If l_startDate >= m_startDate.MinDate _
        AndAlso l_startDate <= m_startDate.MaxDate Then
            m_startDate.Text = l_startDate
        Else
            m_startDate.Text = m_startDate.MinDate
        End If

        Dim l_endDate As Date = l_startDate.AddDays(Period.m_nbDaysInWeek * 3)
        If l_endDate >= m_endDate.MinDate _
        AndAlso l_endDate <= m_endDate.MaxDate Then
            m_endDate.Text = l_endDate
        Else
            m_endDate.Text = m_endDate.MaxDate
        End If

        m_startDate.DateTimeEditor.Value = m_startDate.Text
        m_endDate.DateTimeEditor.Value = m_endDate.Text

        GeneralUtilities.FillWeekTextBox(m_startWeekTB, m_startDate.Text)
        GeneralUtilities.FillWeekTextBox(m_endWeekTB, m_endDate.Text)

    End Sub

#End Region

#Region "Interface"

    Friend Function GetPeriodList() As List(Of Int32)

        'Dim l_weeksIdList As List(Of Int32) = Period.GetWeeksPeriodListFromPeriodsRange(m_startDate.DateTimeEditor.Value.Value, _
        '                                                                                m_endDate.DateTimeEditor.Value.Value)
        'Return Period.GetDaysPeriodsListFromWeeksId(l_weeksIdList)
        Dim l_periodsRange As New List(Of Int32)
        Dim l_date As Date = m_startDate.DateTimeEditor.Value.Value
        Do While l_date <= m_endDate.DateTimeEditor.Value.Value
            l_periodsRange.Add(l_date.ToOADate)
            l_date = l_date.AddDays(1)
        Loop
        Return l_periodsRange

    End Function

    Friend Sub ReinitializePeriodsRange(ByRef p_versionId As UInt32)
        PeriodsRangeSetup(p_versionId)
    End Sub

#End Region

#Region "Events"

    Private Sub m_startDate_ValueChanged(sender As Object, e As EventArgs)
        GeneralUtilities.FillWeekTextBox(m_startWeekTB, m_startDate.DateTimeEditor.Value.Value)
    End Sub

    Private Sub m_endDate_ValueChanged(sender As Object, e As EventArgs)
        GeneralUtilities.FillWeekTextBox(m_endWeekTB, m_endDate.DateTimeEditor.Value.Value)
    End Sub

#End Region

End Class

File: Excel FinancialBI/Excel FinancialBI/MVC/View/PeriodRangeSelectionControl.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace FBI.MVC.View
{
  using Controller;
  using Utils;
  using Model;


  public partial class PeriodRangeSelectionControl : UserControl, IView
  {
    PeriodRangeSelectionController m_controller;
    UInt32 m_versionId;

    public PeriodRangeSelectionControl(UInt32 p_versionId)
    {
      InitializeComponent();
      m_versionId = p_versionId;
    }

    public void LoadView()
    {
      MultilangueSetup();
      m_startDate.FormatValue = "MM-dd-yy";
      m_endDate.FormatValue = "MM-dd-yy";
      PeriodsRangeSetup(m_versionId);
      m_startDate.Validated += Date_ValueChanged;
      m_endDate.Validated += Date_ValueChanged;
    }

    private void MultilangueSetup()
    {
      m_startDateLabel.Text = Local.GetValue("submissionsFollowUp.start_date");
      m_endDateLabel.Text = Local.GetValue("submissionsFollowUp.end_date");
    }

    public void SetController(IController p_controller)
    {
      m_controller = p_controller as PeriodRangeSelectionController;
    }

    private void Date_ValueChanged(object sender, EventArgs e)
    {
      FillWeeksTextbox();
    }

    private void FillWeeksTextbox()
    {
      m_startWeekTB.Text = PeriodModel.GetDateAsStringWeekFormat(m_startDate.DateTimeEditor.Value.Value);
      m_endWeekTB.Text = PeriodModel.GetDateAsStringWeekFormat(m_endDate.DateTimeEditor.Value.Value);
    }

    public void PeriodsRangeSetup(UInt32 p_versionId)
    {
      DateTime l_startDate = m_controller.GetDefaultStartDate();
      DateTime l_endDate = m_controller.GetDefaultEndDate();

      m_startDate.DateTimeEditor.Value = l_startDate;
      m_endDate.DateTimeEditor.Value = l_endDate;

      FillWeeksTextbox();
    }

    public DateTime MinDate
    {
      set
      {
        m_startDate.MinDate = value;
        m_endDate.MinDate = value;
        m_startDate.DateTimeEditor.Value = m_controller.GetDefaultStartDate();
        m_endDate.DateTimeEditor.Value = m_controller.GetDefaultEndDate();
      }
    }

    public DateTime MaxDate
    {
      set
      {
        m_startDate.MaxDate = value;
        m_endDate.MaxDate = value;
        m_startDate.DateTimeEditor.Value = m_controller.GetDefaultStartDate();
        m_endDate.DateTimeEditor.Value = m_controller.GetDefaultEndDate();
      }
    }

    public List<Int32> GetPeriodList()
    {
      return m_controller.GetPeriodRange(m_startDate.DateTimeEditor.Value.Value, m_endDate.DateTimeEditor.Value.Value);
    }


  }
}
